docs/proposals/parquet-storage.md

```diff
@@ -0,0 +1,128 @@
+- Advanced compression techniques reduce storage costs and improve query performance
+- Parallel processing friendly using Parquet Row Group
+
+There are other benefits of Parquet formats, but they are not directly related to the proposal:
+
+- Wide ecosystem and tooling support
+- Column pruning opportunity using projection pushdown
+
+## Out of Scope
+
+- Allow Ingester and Compactor to create Parquet files instead of TSDB blocks directly. This could be in the future roadmap but this proposal only focuses on converting and querying Parquet files.
+
+## Proposed Design
+
+### Components
+
+There are 2 new Cortex components/modules introduced in this design.
+
+#### 1. Parquet Converter
+
+Parquet Converter is a new component that converts TSDB blocks on object store to Parquet file format.
+
+It is similar to compactor, however, it only converts single block. The converted Parquet files will be stored in the same TSDB block folder so that the lifecycle of Parquet file will be managed together with the block.
+
+Only certain blocks can be configured to convert to Parquet file and it will be block duration based, for example we only convert if block duration is >= 12h.
+
+#### 2. Parquet Queryable
+
+Similar to the existing distributorQueryable and blockStorageQueryable, Parquet queryable is a queryable implementation which allows Cortex to query parquet files and can be used in both Cortex Querier and Ruler.
+
+If Parquet queryable is enabled, block storage queryable will be disabled and Cortex querier will not query Store Gateway anymore. But it still queries Ingesters.
+
+Cortex querier remains a stateless component when Parquet queryable is enabled.
+
+### Architecture
+
+```
+┌──────────┐    ┌─────────────┐    ┌──────────────┐
+│ Ingester │───>│   TSDB      │───>│   Parquet    │
+└──────────┘    │   Blocks    │    │  Converter   │
+                └─────────────┘    └──────────────┘
+                                          │
+                                          v
+┌──────────┐    ┌─────────────┐    ┌──────────────┐
+│  Query   │───>│   Parquet   │───>│    Parquet   │
+│ Frontend │    │   Querier   │    │    Files     │
+└──────────┘    └─────────────┘    └──────────────┘
+```
+
+### Data Format
+
+Following the current desgin of Cortex, each Parquet file contains at most 1 day of data.
+
+#### Schema Overview
+
+The Parquet format consists of two types of files:
+
+1. **Labels Parquet File**
+   - Each row represents a unique time series
+   - Each column corresponds to a label name (e.g., `__name__`, `label1`, ..., `labelN`)
+   - Row groups are sorted by `__name__` alphabetically in ascending order
+
+2. **Chunks Parquet File**
+   - Maintains row and row group order matching the Labels file
+   - Contains multiple chunk columns for time-series data. Each column covering a time range of chunks: 0-8h, 8h-16h, 16-24h.
+
+#### Column Specifications
```

Comment:
is this column specification originated by joining the two parquet files above?

maybe I'm missing something obvious, but would be nice to include the rationale for splitting in two files. rows are ordered in the same way both files, so I'm not sure why they are need to be split.

Reply:
Yes you are definitely right. We experimented with a single file with both labels and chunks. The reason of splitting to 2 files is that labels and chunks have kind of different size and read pattern. We are able to configure parquet reader differently for read buffer so that we can read more efficiently.

There is a also [POC from Cloudflare](https://github.com/cloudflare/parquet-tsdb-poc) which uses 2 files so that they can choose to store those files differently. They can cache labels file inmemory and leave chunks file on object store because of size for more efficient index queries.

Overall, 2 files seem a more flexible approach. Maybe @alanprot can share more info.

Reply:
2 Files are useful since the labels parquet file is tiny and can be stored on disk or memoized if wanted. This reduces requests to object storage for any label related lookups.